akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala

```diff
@@ -0,0 +1,299 @@
+/**
+  * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+  */
+
+package akka.http.impl.engine.client
+
+import akka.http.impl.engine.client.RedirectSupportStage.RedirectMapper
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.model.headers.Location
+import akka.http.scaladsl.settings.ClientAutoRedirectSettings
+import akka.http.scaladsl.settings.ClientAutoRedirectSettings.HeadersForwardMode.{All, Only, Zero}
+import akka.stream._
+import akka.stream.scaladsl.BidiFlow
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+
+import scala.collection.immutable
+import scala.language.higherKinds
+import scala.util.control.NoStackTrace
+
+object RedirectSupportStage {
+
+  type RedirectMapper = (ClientAutoRedirectSettings, HttpRequest, HttpResponse) => Option[HttpRequest]
+
+  case class DefaultRedirectMapper() extends RedirectMapper {
+
+    def inferMethod(req: HttpRequest, resp: HttpResponse): Option[HttpMethod] = {
+      import HttpMethods._
+      import StatusCodes._
+      val m = req.method
+      resp.status match {
+        case MovedPermanently | Found | TemporaryRedirect | PermanentRedirect if m == GET || m == HEAD => Some(m)
+        case SeeOther => Some(GET)
+        case _ => None
+      }
+    }
+
+    def sameOrigin(u1: Uri, u2: Uri): Boolean = {
+      u2.isRelative || u1.isAbsolute && u1.authority == u2.authority
+    }
+
+    def apply(s: ClientAutoRedirectSettings, req: HttpRequest, resp: HttpResponse): Option[HttpRequest] = {
+      for {
+        location <- resp.header[Location] if resp.status.isRedirection
```

Comment:
I agree that the pure monadic style looks better. But the suggested change saves the unnecessary location header lookup which will always return `None` for all non-redirect responses and is more expensive to calculate than `isRedirection `.

If only if-statements could come first in for-loops. ;)